python/ray/data/datasource/dbapi2_datasource.py

```diff
@@ -0,0 +1,122 @@
+import math
+from contextlib import contextmanager
+from functools import lru_cache
+from typing import Any, Callable, Generator, Iterable, List, NewType, Optional
+
+import pyarrow as pa
+
+from ray.data._internal.arrow_block import ArrowRow
+from ray.data.block import Block, BlockAccessor, BlockMetadata
+from ray.data.datasource.datasource import Datasource, Reader, ReadTask, WriteResult
+
+# A Python DB API2-compliant `Connection` object.
+Connection = Any
+
+
+def cursor_to_block(cursor) -> Block:
+    rows = cursor.fetchall()
+    columns = [column_description[0] for column_description in cursor.description]
+    pylist = [{c: v for c, v in zip(columns, row)} for row in rows]
+    return pa.Table.from_pylist(pylist)
+
+
+class DBAPI2Datasource(Datasource[ArrowRow]):
+    def __init__(self, connection_factory: Callable[[], Connection]):
+        self.connection_factory = connection_factory
+
+    def create_reader(self, sql: str) -> "Reader":
+        return DBAPI2Reader(sql, self.connection_factory)
+
+
+@contextmanager
+def connect(
+    connection_factory: Callable[[], Connection]
+) -> Generator[Connection, None, None]:
+    connection = connection_factory()
+    try:
+        yield connection
+    except:
+        # `rollback` is optional since not all databases provide transaction support.
+        if hasattr(connection, "rollback"):
+            connection.rollback()
+    finally:
+        connection.close()
+
+
+class DBAPI2Reader(Reader):
+
+    NUM_SAMPLE_ROWS = 100
+
+    def __init__(self, sql: str, connection_factory: Callable[[], Connection]):
+        self.sql = sql
+        self.connection_factory = connection_factory
+
+    @property
+    @lru_cache
```

Comment:
for my own education - why we need this `lru_cache` annotation? Are we caching the result somehow?

Reply:
`lru_cache` is to prevent re-executing queries to count the number of rows or retrieve a sample row. It's a decorator provided by the built-in `functools` package.

Turns out there's a memory leak bug when you use `lru_cache` with methods, so I've removed it. (Eric uses `cached_property` which doesn't leak memory, but that isn't available until Python 3.8).